### PR TITLE
[WIP] Template Loader

### PR DESCRIPTION
**Pull request by vicb**
Work in progress for the Template Loader.

What is implemented:
- `TemplateLoader.load()` returns a Promise & `Compiler.compile()` is async,
- The shadow DOM strategy is now global to the application (vs component specific before)

TODO:
- [x] extract the styles from the templates when the emulated strategy is used (insert them in the `<head>`),
- [x] html import based template loader

### CODE REVIEW

**Comment by vsavkin** [comment]
I am probably missing something, but it looks like you are making the strategy global. This means that you won't be able to use emulated and shadow-dom components in the same application. I think we need this capability.

**Comment by vicb** [comment]
Victor, that's right.
Having a global strategy is what we have discussed with Misko and Tobias yesterday.
Why do you think we need to mix both strategies ?

File: modules/core/src/application.js [hunk 1]
```diff
@@ -52,12 +56,21 @@ function _injectorBindings(appComponentType) {
         return element;
       }, [appComponentAnnotatedTypeToken, appDocumentToken]),
 
-      bind(appViewToken).toAsyncFactory((changeDetection, compiler, injector, appElement,
-            appComponentAnnotatedType) => {
+      bind(appViewToken).toAsyncFactory((changeDetection,
```
Comment (rkirov) [inline modules/core/src/application.js]: revert back to argument-packing indentation style. We have to use consistent style in the codebase and argument-packing vs new-line style is preferred by most teammembers.

I don't want to start a big debate on the merits of each, but we have to be consistent.

File: modules/core/src/compiler/compiler.js [hunk 2]
```diff
@@ -53,59 +50,104 @@ export class Compiler {
...
+          ListWrapper.push(componentPromises, componentPromise);
+        }
       }
-    }
 
-    return rootProtoView;
+      // The protoView is resolved after all the components in the template
+      // have been compiled.
+      return PromiseWrapper
+        .all(componentPromises)
```
Comment (rkirov) [inline modules/core/src/compiler/compiler.js]: How does `all` handle error handling. Would this promise reject if one of its dependents fail (not all)?

File: modules/core/src/compiler/compiler.js [hunk 3]
```diff
@@ -53,59 +50,104 @@ export class Compiler {
...
+        .then(function(_) {
+          return protoView;
+        })
+    });
+
+    MapWrapper.set(this._compiling, cmpMetadata.type, pvPromise);
+
+    return pvPromise;
+  }
+
```
Comment (rkirov) [inline modules/core/src/compiler/compiler.js]: What happens when a `tmpLoader` fails? We might need to attach `pvPromise.catch()` callback to have a more descriptive erorr.

File: modules/core/src/compiler/template_loader.js [hunk 4]
```diff
@@ -1,11 +1,52 @@
...
+import {DirectiveMetadata} from './directive_metadata';
 //import {Document} from 'facade/dom';
 
 /**
  * Strategy to load component templates.
  */
 export class TemplateLoader {
-  load(url:string):Promise<Document> {
-    return null;
+  // todo(vicb): was Promise<Document>
```
Comment (rkirov) [inline modules/core/src/compiler/template_loader.js]: change this `todo` to something more descriptive. It doesn't tell me what you have `to do` :)

File: modules/core/src/compiler/template_loader.js [hunk 5]
```diff
@@ -1,11 +1,52 @@
...
+
+    if (isPresent(tplConfig.inline)) {
+      var template = DOM.createTemplate(tplConfig.inline);
+      return PromiseWrapper.resolve(template);
+    }
+
+    if (isPresent(tplConfig.url)) {
+      var url = tplConfig.url;
+      var completer = PromiseWrapper.completer();
+      var link = DOM.createElement('link');
```
Comment (rkirov) [inline modules/core/src/compiler/template_loader.js]: What would be the alternative to html5 imports? According to http://caniuse.com/#search=imports we going to miss evergreen Firefox and iOS Safari. Add a comment to remind us what to do.

File: modules/core/src/compiler/template_loader.js [hunk 6]
```diff
@@ -1,11 +1,52 @@
...
+      link.rel = 'import';
+      link.href = url;
+      DOM.on(link, 'load', function() {
```
Comment (rkirov) [inline modules/core/src/compiler/template_loader.js]: use fat errors `=> ()` ;)

File: modules/core/test/compiler/compiler_spec.js [hunk 7]
```diff
@@ -92,9 +94,27 @@ export function main() {
...
+          current.inheritedElementBinder = current.inheritedProtoView.bindElement(null);
+          current.componentDirective = reader.read(NestedComponent);
+          ListWrapper.push(nestedElBinders, current.inheritedElementBinder);
+        }
+      });
+      compiler.compile(MainComponent, mainEl).then( (protoView) => {
+        expect(nestedElBinders[0].nestedProtoView).toBe(nestedElBinders[1].nestedProtoView);
+        done();
+      });
+
```
Comment (rkirov) [inline modules/core/test/compiler/compiler_spec.js]: remove extra newline.

File: modules/benchmarks/src/compiler/compiler_benchmark.js [hunk 8]
```diff
@@ -89,14 +96,14 @@ export function main() {
     // Need to clone every time as the compiler might modify the template!
     var cloned = DOM.clone(templateNoBindings);
     cache.clear();
-    compiler.compileAllLoaded(null, annotatedComponent, cloned);
+    compiler.compile(BenchmarkComponent, cloned);
```
Comment (rkirov) [inline modules/benchmarks/src/compiler/compiler_benchmark.js]: I think @tbosch used this to skip the promise path for benchmarks. I am ok with removing it given the extra cost of keeping the fast pass when templateLoader is there too. So just a heads up to Tobias.

**Comment by mhevery** [comment]
- [ ] I don't see tests which shows how compiler loads the templates. I would expect to see some kind of mock.
- [ ] We talked about switching to XHR from `<link>`
- [ ]  We need to have a way to mock out the Template loading XHRs.

File: modules/benchmarks/src/compiler/compiler_benchmark.js (hunk 8 above)
Comment (vicb) [inline modules/benchmarks/src/compiler/compiler_benchmark.js]: Compiler.compile is now always async, there is no more sync path.

File: modules/core/src/compiler/template_loader.js (hunk 4 above)
Comment (vicb) [inline modules/core/src/compiler/template_loader.js]: I'll update this.
We can only use a `Document` with html import and the preferred way for now will be to use XHR

File: modules/core/src/compiler/template_loader.js (hunk 5 above)
Comment (vicb) [inline modules/core/src/compiler/template_loader.js]: As discussed with Tobias and Misko, XHR will be preferred for now

File: modules/core/src/compiler/template_loader.js (hunk 6 above)
Comment (vicb) [inline modules/core/src/compiler/template_loader.js]: Seems like (from a PR comment) `function` is faster when you don't need to bind this (probably applicable to ES5 only)

File: modules/core/test/compiler/compiler_spec.js (hunk 7 above)
Comment (vicb) [inline modules/core/test/compiler/compiler_spec.js]: consistency...

File: modules/core/src/compiler/compiler.js (hunk 2 above)
Comment (vicb) [inline modules/core/src/compiler/compiler.js]: It fails if any of the dependents fails

File: modules/core/src/compiler/compiler.js (hunk 3 above)
Comment (vicb) [inline modules/core/src/compiler/compiler.js]: I plan to add more descriptive errors

**Comment by rkirov** [comment]
merged (after a monster of a rebase :))

**Comment by vicb** [comment]
@rkirov thanks !

**Pull request closed without merging**